feat: Improvements to HOL blocking article (#40184)
mdn/content: files/en-us/glossary/head_of_line_blocking/index.md

---
title: Head-of-line blocking
short-title: HOL blocking
slug: Glossary/Head_of_line_blocking
page-type: glossary-definition
sidebar: glossarysidebar
---

In computer networking, **head-of-line blocking** (_HOL blocking_) refers to a performance bottleneck that occurs when a queue of packets is held up by the first packet in the queue, even though other packets in the queue could be processed.

> In HTTP/1.1, HOL blocking occurs when a client sends multiple requests to a {{glossary("server")}} over a single, persistent TCP connection, and one of the responses is delayed for any reason — such as network congestion, {{glossary("TCP slow start")}}, or, problems in transit.

Author: ```suggestion
In HTTP/1.1, HOL blocking occurs when a client sends multiple requests to a {{glossary("server")}} per TCP connection, and a response is delayed for any reason — such as network congestion, {{glossary("TCP slow start")}}, or, problems in transit.
```

The browser will use about 6 TCP connections per origin, so we want to say when one of these gets used for multiple requests, basically.

Author: Maybe there's a better reword of that

Reviewer: I'm prefer the original, but how about a compromise (the per TCP does not work for me).

```suggestion
In HTTP/1.1, HOL blocking occurs when a client sends multiple requests to a {{glossary("server")}} over a particular TCP connection, and a response on that connection is delayed for any reason — such as network congestion, {{glossary("TCP slow start")}}, or, problems in transit.
```

Author: Good, let's go with that 👍🏻

> HTTP/1.1 requests are sent sequentially, so a delay in fully receiving a response blocks the next request-response exchanges.

Author: ```suggestion
HTTP/1.1 requests are sent sequentially per TCP connection, so a delay in fully receiving a response blocks the next request-response exchange.
```

In computer networking, **head-of-line blocking** (_HOL blocking_) refers to a performance bottleneck that occurs when a queue of packets is held up by the first packet in the queue, even though other packets in the queue could be processed.

> In HTTP/1.1, HOL blocking occurs when a client sends multiple requests to a {{glossary("server")}} over a single, persistent TCP connection.

Reviewer: So it _always_ occurs if you sent multiple requests in this case?

A response can be delayed due to network congestion, {{glossary("TCP slow start")}}, problems in transit, or other reasons.
HTTP/1.1 requests are sent sequentially, so a delay in fully receiving a response blocks the next request-response exchanges.
A mechanism called _HTTP pipelining_ tried to work around this, where multiple requests were sent off by a client without waiting for any responses.
Pipelining proved tricky to implement in reality, so this mechanism is rarely used, if ever, and most browsers no longer support it.

> HTTP/2 addresses some problems in HTTP/1.1 through request _multiplexing_.

Author: ```suggestion
HTTP/2 addresses the HOL blocking problems in HTTP/1.1 through request _multiplexing_.
```

> In HTTP/1.1, HOL blocking occurs when a client sends multiple requests to a {{glossary("server")}} over a single, persistent TCP connection.
> A response can be delayed due to network congestion, {{glossary("TCP slow start")}}, problems in transit, or other reasons.
> HTTP/1.1 requests are sent sequentially, so a delay in fully receiving a response blocks the next request-response exchanges.
> A mechanism called _HTTP pipelining_ tried to work around this, where multiple requests were sent off by a client without waiting for any responses.
> Pipelining proved tricky to implement in reality, so this mechanism is rarely used, if ever, and most browsers no longer support it.
> 
> HTTP/2 addresses the HOL blocking problems in HTTP/1.1 through request _multiplexing_.
> Multiplexing means a single TCP connection can be used to interleave requests and responses in numbered streams.
> A client can send many requests over a single connection without waiting for earlier responses.
> This eliminates HOL blocking in the protocol layer, but it still exists at the transport ({{glossary("TCP")}}) layer.

Reviewer: The first paragraph, in particular felt clunky. How about this (or some variant)

```suggestion
In HTTP/1.1, HOL blocking occurs when a client sends multiple requests to a {{glossary("server")}} over a single, persistent TCP connection, and one of the responses is delayed for any reason — such as network congestion, {{glossary("TCP slow start")}}, or, problems in transit.
HTTP/1.1 requests are sent sequentially, so a delay in fully receiving a response blocks the next request-response exchanges.

A mechanism called _HTTP pipelining_ tried to work around this, where multiple requests were sent off by a client without waiting for any responses.
Pipelining proved tricky to implement in reality, so this mechanism is rarely used, if ever, and most browsers no longer support it.

HTTP/2 addresses the HOL blocking problems in HTTP/1.1 through request _multiplexing_.
Multiplexing allows a single TCP connection to interleave requests and responses in numbered streams.
A client can send many requests over a single connection without waiting for earlier responses.
Note that while HOL has been fixed in HTTP, it is still a problem at the transport ({{glossary("TCP")}}) layer level.
```
